### DISCUSSION TITLE
@when to work dynamically

### DISCUSSION THREAD

**Discussion by Neon22** [pyscript/pyscript#1845] [Proposals]
`@when` does not work, and is not intended to work, with dynamic changes to the DOM.
It runs once, hooks up event listeners to the DOM objects that are selected, and that's it.
It's a fairly thin wrapper around around `EventListener`. That's all it's meant to do.

IWBNI it could deal with dynamic changes to the dom. There are several ways to implement this. Some more pythonic than others.

As a straw man I propose this simple method which involves minimal code change and is, perhaps, mildly inefficient and not reversible.

**Proposal:**
When the @when decorator runs it does not add event listeners directly.
Instead it makes a dictionary on the JS side which holds:
 - the dom element selector,
 - event/s,
 - target function (with args?)

This is held in the 'MutationObserver'  which provides the ability to watch for changes being made to the DOM.

- Added code in there checks new dom entries against the selector and adds event listeners dynamically.
- (Also responsible for removing them before removing dom elements to help with mem leaks - which is what I believe the pyiodide one does currently. Perhaps it can be used as a starting point).

The advantage of doing it this way is having the pythonic decorator work on dynamic dom changes.
The disadvantage is the overhead of checking each dom change against the registered selectors.  Potentially a problem but I suspect will not be in practice. Also more functions would need to be added to remove an event listener from an element. However this may not be needed very often in practice.

The above approach suggested only as a point of initial discussion.

**Comment by WebReflection** [comment]
I have a library for that: https://github.com/WebReflection/qsa-observer#readme
it's used by the only Custom Elements polyfill that patches builtin extends too and it's been battle-tested for ages.

The orchestration is simplified, but there's still some orchestration to handle ... specially with worker case ... still doable to me, as it's possible to add specific selectors and whenever a node is live or disconnected we can add/remove the related listener.

That being said, I don't have a strong opinion about this, not sure it's super needed, yet I'd be happy to help forward if we decide to go this path.

/cc @JeffersGlass @fpliger

**Comment by WebReflection** [comment]
P.S. nodes removed don't leak listeners ... listeners work like in a WeakMap, when the node is gone, the GC frees attached listeners too.

**Reply by Neon22** [reply]
Sorry. Its my misinterpretation of create-proxy in pyiodide. which is instead perhaps about GC on the python side. I vaguely recall a side discussion about needing something similar to enable micropython to do "UI stuff" as well as pyioidide.
- https://jeff.glass/post/pyscript-why-create-proxy/

**Reply by JeffersGlass** [reply]
The `PyProxy` objects (that is, JS objects which proxy Python objects) are garbage collected [if the browser supports FinalizationRegistry](https://github.com/pyodide/pyodide/pull/1306) and if there are no Python => JS => Python reference cycles. So that's accounted for, so long as we `destroy()` any proxies that aren't actively being used, I suppose. (I'm making some assumptions about how the orchestration would work here.)

PS. That PR has a hilarious note: _"If the FinalizationRegistry API is missing... it gracefully falls back to leaking objects like crazy."_ Thankfully, it seems to [fairly widely supported](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry).

**Comment by JeffersGlass** [comment]
I think there's value in having this feature - clearly, at least @Neon22 would find it quite useful. 

> The advantage of doing it this way is having the pythonic decorator work on dynamic dom changes

I think this is a different feature from `@when`, which is basically "addEventListener as a decorator." I think we should have a _new_ feature, with a new name, that accomplishes this, and leave `@when` as it is.

What we call this, I don't know. The one that's stuck in my head is `@whenever`, but I think that's trying to be too cute, and would be confusing. Maybe `@register_dynamic_listener`? Quite wordy, that...

**Reply by WebReflection** [reply]
am I being too silly to even think about a `@whenever` decorator, (*edit*) like you did? 😅

**Reply by WebReflection** [reply]
seriously though, the JS ecosystem had `on(...args)` and `once(...args)` around for ages, and I don't remember anyone ever complaining about any of the two, "*risonance*" wise ... I think `@whenever` is likely the best fit / word for the job, as I can see already ourselves describing the feature as "*whenever that selector lands on the page, the listener will be attached to it*", no matter which name we pick.

**Reply by fpliger** [reply]
+1 on being a new callable slightly differen then `@when`

`@whenever` is nice :)  and I think would be my preference... I don't dislike `on` (althought... would that be confusing in terms of `JS` vs `Python` ecosystems) or `@observe` either

**Comment by Neon22** [comment]
I'll just put this here: - [Mutation Summary](https://github.com/rafaelw/mutation-summary/blob/master/APIReference.md#supported-selector-syntax)
has limited selector capability and/but runs on the JS side.

**Reply by WebReflection** [reply]
that's 8 years ago and heavier than `qsa-observer` + I maintain the polyfill for CE and the library so ... whatever works is fine but then again, I maintain that lib and I can guarantee it works from any Desktop to any Mobile browser ;-)
